### PR TITLE
Modifies ia_sync API for hank

### LINKED ISSUE #7539: Create endpoint for Archive.org to update OCAIDs in OL records
### Describe the problem that you'd like solved

Today, there is no way for Internet Archive to notify Open Library of important changes to records.  If data becomes out-of-sync between the two platforms, Open Library records must be updated and synced by a human.  Before this is done, somebody has to first notice that the data is out-of-sync.  This is not ideal...

### Proposal & Constraints

Edited proposal:
Create a new POST handler and endpoint for syncing OCAID changes.  Sync requests must contain S3 credentials, edition `olid`, the original `ocaid`, and either `is_dark` if the item has been made dark, or `new_ocaid` if the OCAID has been changed.

### Stakeholders

@mekarpeles

### PR DESCRIPTION
**Pull request by mekarpeles**
Closes #7539

### CODE REVIEW

File: openlibrary/plugins/admin/code.py [hunk 1]
```diff
@@ -240,74 +241,71 @@ def GET(self):
...
-        All requests must have an olid and an action.  If the action is
-        'modify', the request must also include 'ocaid'.
-        """
-        action = i.get('action', '')
-        return 'olid' in i and (
-            action == 'remove' or (action == 'modify' and 'ocaid' in i)
-        )
+    def get_edition_by_ocaid(self, ocaid):
+        ocaid = ocaid.replace('_', ' ')
+        q = {'type': '/type/edition', 'ocaid': ocaid}
```
Comment (jimchamp) [inline openlibrary/plugins/admin/code.py:296]: I modeled this on the code used to display a book page when an `ocaid` is used, [here](https://github.com/internetarchive/openlibrary/blob/409914bf541b32b2160200b7623060f2b5fab6c0/openlibrary/plugins/openlibrary/code.py#L500-L513).  Not sure if the underscore replacement is still necessary these days.

File: openlibrary/plugins/admin/code.py [hunk 2]
```diff
@@ -240,74 +241,73 @@ def GET(self):
...
 
-        if not auth.get('username', ''):
-            return False
+        if auth.get('username', ''):
+            acct = accounts.OpenLibraryAccount.get(email=auth.get('username'))
+            user = acct.get_user() if acct else None
 
-        acct = accounts.OpenLibraryAccount.get(email=auth.get('username'))
-        user = acct.get_user() if acct else None
+            if user and user.is_usergroup_member('/usergroup/ia'):
```
Comment (jimchamp) [inline openlibrary/plugins/admin/code.py:289]: TODO: Populate usergroup.

File: openlibrary/plugins/admin/code.py [hunk 3]
```diff
@@ -240,74 +241,73 @@ def GET(self):
...
+            raise web.badrequest(
+                f'Failed to lookup archive.org item with ocaid {ocaid}'
+            )
+        if not metadata.get('openlibrary_edition'):
+            edition = self.get_edition_by_ocaid(i.old_ocaid or ocaid)
+            if not edition:
+                raise web.badrequest('No openlibrary_edition specified in item')
 
+        if not edition:
+            edition = web.ctx.site.get(f'/books/{metadata.get("openlibrary_edition")}')
```
Comment (mekarpeles) [inline openlibrary/plugins/admin/code.py:262]: Can we simplify this to...

```
if metadata.get('openlibrary_edition'):
    edition = web.ctx.site.get(f'/books/{metadata.get("openlibrary_edition")}')
else:
    edition = self.get_edition_by_ocaid(i.old_ocaid or ocaid)
if not edition:
    raise web.badrequest('No openlibrary_edition specified in item')
```

File: openlibrary/plugins/admin/code.py [hunk 4]
```diff
@@ -240,74 +241,73 @@ def GET(self):
...
-        action = i.get('action', '')
-        return 'olid' in i and (
-            action == 'remove' or (action == 'modify' and 'ocaid' in i)
-        )
+    def get_edition_by_ocaid(self, ocaid):
+        ocaid = ocaid.replace('_', ' ')
+        q = {'type': '/type/edition', 'ocaid': ocaid}
+        return web.ctx.site.things(q)
 
     def remove_ocaid(self, edition):
```
Comment (mekarpeles) [inline openlibrary/plugins/admin/code.py:299]:
```suggestion
    def remove_ocaid(self, edition, msg=""):
```

File: openlibrary/plugins/admin/code.py [hunk 5]
```diff
@@ -240,74 +241,73 @@ def GET(self):
...
-        """Deletes OCAID from given edition"""
+        """Deletes OCAID from given edition.  Does not affect
+        entries in `source_records` list.
+        """
         data = edition.dict()
         del data['ocaid']
         web.ctx.site.save(data, 'Remove OCAID: Item no longer available to borrow.')
```
Comment (mekarpeles) [inline openlibrary/plugins/admin/code.py:305]:
```suggestion
        comment = ' '.join([i for i in ['Remove OCAID', msg] if i])
        web.ctx.site.save(data, comment)
```

File: openlibrary/plugins/admin/code.py (hunk 2 above)
Comment (mekarpeles) [reply openlibrary/plugins/admin/code.py:289]: This is likely a case where we can delegate directly to a role account on Archive.org an no Open Library account is required (e.g. use xauthn to ensure that the owner of these S3 keys is e.g. `@iasync` role account)

You should be able to call the existing xauth or s3auth functions:
https://github.com/internetarchive/openlibrary/blob/master/openlibrary/accounts/model.py#L711C7-L729

Comment (mekarpeles) [reply openlibrary/plugins/admin/code.py:289]: Alternatively, `admin` likely good enough for now, you decide.

File: openlibrary/plugins/admin/code.py (hunk 1 above)
Comment (jimchamp) [reply openlibrary/plugins/admin/code.py:296]:
```suggestion
        q = {'type': '/type/edition', 'ocaid': ocaid}
```

**Comment by cdrini** [comment]
Taking of testing to test another PR ( #8550 )

**Pull request closed without merging**